raster-foundry/raster-foundry
# Normalize pagination and add to all group object lists

## Overview

This PR accomplishes a few things: adds a method for automatically handling angular dependency injection assignment, adds a pagination service to handle common pagination patterns, and adds pagination to all currently available object list pages.

#### Automatic Dependency Injection Assignment

Previously, each dependency declared in a class had to be assigned to a matching property on the class:

```js
class Controller {
    constructor($scope, $stateParams, modalService, organizationService) {
        'ngInject';
        this.$scope = $scope;
        this.$stateParams = $stateParams;
        this.modalService = modalService;
        this.organizationService = organizationService;
    }
}
```

With `autoInject`, we can now save ourselves a good bit of boilerplate:

```js
class Controller {
    constructor($scope, $stateParams, modalService, organizationService) {
        'ngInject';
        $scope.autoInject(this, arguments);
    }
}
```

The method is defined on the `$rootScope`, so it is available anywhere `$scope` is accessible (not services). After the call to `autoInject`, each argument is assigned to a matching property on the class (`this.$stateParams` or `this.organizationService`).

#### Pagination Service

As of this PR the pagination service has three methods for common pagination tasks: 

- `buildPagedSearch(context)` returns a debounced method that calls the passed context's `fetchPage` method
- `buildPagination(paginatedResponse)` returns an object with the commonly used pagination values
- `updatePageParam(page)` updates the current URL if necessary to include a `page` parameter

#### Paginated Object List Pages

The following pages have had the pagination refactored or added:

- [x] platform/projects
- [x] platform/rasters
- [x] platform/users
- [x] platform/organizations
- [x] organization/projects
- [x] organization/rasters
- [x] organization/users
- [x] organization/teams
- [x] team/projects
- [x] team/rasters
- [x] team/users
- [x] user/projects
- [x] user/rasters
~- [ ] user/organizations~
~- [ ] user/teams~

### Checklist


## Testing Instructions

 For each of the pages:
  * make sure a list is rendered when there is an applicable item
  * make sure pagination controls only appear if the page size is exceeded (you may benefit from setting small page sizes in the requests so that you don't need to share a lot of things)
  * for the pages with more advanced capabilities (inviting users, etc), test these functions and make sure they still work

Closes #3612
